swiftide-indexing/src/loaders/file_loader.rs

```diff
@@ -5,118 +5,181 @@ use std::{
-                let content = String::from_utf8_lossy(&buf);
-
-                let original_size = content.len();
-
-                Node::builder()
-                    .path(entry.path())
-                    .chunk(content)
-                    .original_size(original_size)
-                    .build()
-            })
+        Iter::new(&self.root, self.extensions.clone()).fuse()
     }
 }
 
-// Helper function to check if a file has the specified extension.
-// If no extensions are specified, this function will return true.
-// If the file has no extension, this function will return false.
-fn file_has_extension(extensions: Option<&[impl AsRef<str>]>, path: &Path) -> bool {
-    extensions.as_ref().is_none_or(|exts| {
-        let Some(ext) = path.extension() else {
-            return false;
-        };
-        exts.iter()
-            .any(|e| e.as_ref() == ext.to_string_lossy().as_ref())
-    })
+/// An iterator that walks over the files in a directory and loads them.
+///
+/// This is a private struct that is used to implement the `FileLoader` iterator.
+struct Iter {
+    /// The walk instance that iterates over the files in the directory.
+    walk: Walk,
+    /// The extensions to include.
+    include_extensions: Option<Vec<String>>,
+    /// A span that tracks the current file loader.
+    span: Span,
+}
+
+impl Iterator for Iter {
+    type Item = anyhow::Result<Node>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let _span = self.span.enter();
+        loop {
+            // stop the iteration if there are no more entries
+            let Some(entry) = self.walk.next() else {
+                return None;
+            };
+
+            // propagate any errors that occur during the directory traversal
+            let entry = match entry {
+                Ok(entry) => entry,
+                Err(err) => return Some(Err(err.into())),
+            };
+
+            if let Some(node) = self.load(entry) {
+                return Some(node);
+            }
+        }
+    }
+}
+
+impl Iter {
+    /// Creates a new `Iter` instance.
+    fn new(root: &Path, include_extensions: Option<Vec<String>>) -> Self {
+        let span = debug_span!("file_loader", root = %root.display());
+        tracing::debug!(parent: &span, extensions = ?include_extensions, "Loading files");
+        Self {
+            walk: Walk::new(root),
+            include_extensions,
+            span,
+        }
+    }
+
+    #[instrument(skip_all, fields(path = %entry.path().display()))]
+    fn load(&self, entry: DirEntry) -> Option<anyhow::Result<Node>> {
+        if entry.file_type().is_some_and(|ft| !ft.is_file()) {
+            // Skip directories and non-files
+            return None;
+        }
+        if let Some(extensions) = &self.include_extensions {
+            let Some(extension) = entry.path().extension() else {
+                tracing::trace!("Skipping file without extension");
```

Comment:
These are verbose, but the sort of verbose that you enable because there's some stupid problem you're bashing your head against in your code which is immediately obvious when turned on.